src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceProgress.cs

```diff
@@ -60,19 +70,56 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
                 _pendingProgress = new PendingProgress();
             }
 
-            _pendingProgress.Update(sourceId, record);
-
             if (_progPane == null)
             {
-                // This is the first time we've received a progress record. Create a progress pane, then show it.
+                // This is the first time we've received a progress record
+                // Create a progress pane
+                // Set up a update flag
+                // Create a timer for updating the flag
 
                 _progPane = new ProgressPane(this);
+
+                if (_progPaneUpdateTimer == null && _progPane != null)
+                {
+                    // Show a progress pane at the first time we've received a progress record
+                    progPaneUpdateFlag = true;
+
+                    // Invoke the timer only once to exclude overlaps
+                    // The timer will be restarted in 'ProgressPaneUpdateTimerElapsed'
+                    _progPaneUpdateTimer = new Timer( new TimerCallback(ProgressPaneUpdateTimerElapsed), null, UpdateTimerThreshold, Timeout.Infinite);
+                }
+            }
+
+            if (progPaneUpdateFlag || record.RecordType == ProgressRecordType.Completed)
+            {
+                // Update the progress pane only when the timer set up the update flag or WriteProgress is completed.
+                // As a result, we do not block WriteProgress and whole script and eliminate unnecessary console locks and updates.
+                _pendingProgress.Update(sourceId, record);
+                _progPane.Show(_pendingProgress);
+                progPaneUpdateFlag = false;
             }
-            _progPane.Show(_pendingProgress);
         }
 
 
 
+        /// <summary>
+        ///
+        /// TimerCallback for _progPaneUpdateTimer to update 'progPaneUpdateFlag' and restart the timer
+        ///
+        /// </summary>
+
+        private
+        void
+        ProgressPaneUpdateTimerElapsed(object sender)
+        {
+            if (_progPane != null)
```

Comment:
If we make the timer repeatedly fire by itself, then we can avoid the call to `timer.Change`. And it actually doesn't matter whether `_progPane` is null or not because the worst that can happen is that the timer set the flag to 1 and we call `show` one more time.

Given that, I think this method can be simplified to this:
```
        private
        void
        ProgressPaneUpdateTimerElapsed(object sender)
        {
            Interlocked.CompareExchange(ref progPaneUpdateFlag, 1, 0);
        }
```